src/ethereum_spec_tools/evm_tools/t8n/env.py

```diff
@@ -114,15 +113,46 @@ def read_excess_blob_gas(self, data: Any, t8n: "T8N") -> None:
                 data["parentBlobGasUsed"], U64
             )
 
-        excess_blob_gas = (
+        if self.excess_blob_gas is not None:
+            return
+
+        assert self.parent_excess_blob_gas is not None
+        assert self.parent_blob_gas_used is not None
+
+        parent_blob_gas = (
             self.parent_excess_blob_gas + self.parent_blob_gas_used
         )
 
         target_blob_gas_per_block = t8n.fork.TARGET_BLOB_GAS_PER_BLOCK
 
-        self.excess_blob_gas = U64(0)
-        if excess_blob_gas >= target_blob_gas_per_block:
-            self.excess_blob_gas = excess_blob_gas - target_blob_gas_per_block
+        if parent_blob_gas < target_blob_gas_per_block:
+            self.excess_blob_gas = U64(0)
+            return
+
+        if t8n.fork.is_after_fork("ethereum.osaka"):
+            assert self.parent_base_fee_per_gas is not None
+
+            GAS_PER_BLOB = t8n.fork.GAS_PER_BLOB
+            BLOB_BASE_COST = t8n.fork.BLOB_BASE_COST
+            BLOB_SCHEDULE_MAX = t8n.fork.BLOB_SCHEDULE_MAX
+            BLOB_SCHEDULE_TARGET = t8n.fork.BLOB_SCHEDULE_TARGET
+
+            target_blob_gas_price = Uint(GAS_PER_BLOB)
+            target_blob_gas_price *= t8n.fork.calculate_blob_gas_price(
+                self.parent_excess_blob_gas
+            )
+
+            base_blob_tx_price = BLOB_BASE_COST * self.parent_base_fee_per_gas
+            if base_blob_tx_price > target_blob_gas_price:
+                blob_schedule_delta = BLOB_SCHEDULE_MAX - BLOB_SCHEDULE_TARGET
+                self.excess_blob_gas = (
+                    self.parent_excess_blob_gas
+                    + self.parent_blob_gas_used
+                    * blob_schedule_delta
+                    // BLOB_SCHEDULE_MAX
+                )
```

Comment:
We get none for `excess_blob_gas` without this (i.e when not triggering the reserve)
```suggestion
                )
           else:
               self.excess_blob_gas = parent_blob_gas - target_blob_gas_per_block
```

Instead of the latter, I think it could be better to have the default 4844 first, setting `excess_blob_gas = parent - target`. Then breaking down into the `if` to overwrite with the 7918 behaviour if the reserve is active etc.

Reply:
Updated